File: samples_c/s109_apache_getword.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

/* Kaynak: Apache HTTP Server server/util.c, ap_getword() - birebir mantik
   (apr_pool_t/apr_pstrmemdup soyutlamalari, havuzsuz duz malloc+strncpy ile
   degistirildi; islevsel davranis AYNI).
   https://raw.githubusercontent.com/apache/httpd/trunk/server/util.c
   Lisans: Apache-2.0.
   Girdi: iki satir - durdurma karakteri (tek karakter) ve ayristirilacak
   metin.
   Cikti: metindeki ARDISIK "kelime" alanlari, her biri kendi satirinda,
   metin tukenene kadar tekrar tekrar cikarilir (gercek ap_getword'un
   cagri-donguleriyle ayni sekilde). */

char *my_getword(const char **line, char stop) {
    const char *pos = *line;
    int len;
    char *res;

    while ((*pos != stop) && *pos) {
        ++pos;
    }

    len = (int)(pos - *line);
    res = malloc(len + 1);
    memcpy(res, *line, len);
    res[len] = 0;

    if (stop) {
        while (*pos == stop) {
            ++pos;
        }
    }
    *line = pos;

    return res;
}

int main(void) {
    char stopc[8], text[512];
    if (!fgets(stopc, sizeof(stopc), stdin)) return 0;
    if (!fgets(text, sizeof(text), stdin)) return 0;
    stopc[strcspn(stopc, "\r\n")] = 0;
    text[strcspn(text, "\r\n")] = 0;
    if (stopc[0] == 0) return 0;

    const char *p = text;
    while (*p) {
        char *w = my_getword(&p, stopc[0]);
        printf("%s\n", w);
        free(w);
    }
    return 0;
}

File: translations_rust__gemini/s109_apache_getword.rs
use std::io::{self, BufRead};

fn my_getword<'a>(line: &mut &'a [u8], stop: u8) -> &'a [u8] {
    let mut pos = 0;
    while pos < line.len() && line[pos] != stop {
        pos += 1;
    }

    let res = &line[..pos];

    if stop != 0 {
        while pos < line.len() && line[pos] == stop {
            pos += 1;
        }
    }

    *line = &line[pos..];
    res
}

fn main() {
    let stdin = io::stdin();
    let mut handle = stdin.lock();

    let mut stopc = Vec::new();
    if handle.read_until(b'\n', &mut stopc).unwrap_or(0) == 0 {
        return;
    }

    let mut text = Vec::new();
    if handle.read_until(b'\n', &mut text).unwrap_or(0) == 0 {
        return;
    }

    if let Some(pos) = stopc.iter().position(|&b| b == b'\r' || b == b'\n') {
        stopc.truncate(pos);
    }
    if let Some(pos) = text.iter().position(|&b| b == b'\r' || b == b'\n') {
        text.truncate(pos);
    }

    if stopc.is_empty() {
        return;
    }

    let stop = stopc[0];
    let mut p = &text[..];

    while !p.is_empty() {
        let w = my_getword(&mut p, stop);
        println!("{}", String::from_utf8_lossy(w));
    }
}
